Some of the queues are closed

lavinmq version: 1.2.5

The status of some queues in the service becomes closed.
I didn't find any documentation about the queues being closed, is there a way to change them to a running status?

> What does the log say?

```log
2024-04-26 10:12:31 +00:00 [ERROR] queue[vhost=/ name=icp_inc_queue] - Queue closed due to error
2024-04-26 10:12:31 +00:00 [ERROR] client[vhost=/ address=10.42.0.160:42596] - Unexpected error, while reading: Channel is closed (LavinMQ::Queue::MessageStore::ClosedError)
  from /usr/share/crystal/src/hash.cr:1080:49 in 'delete'
  from /usr/share/crystal/src/mutex.cr:136:7 in 'delete_message'
  from /tmp/src/lavinmq/queue/queue.cr:750:7 in 'ack'
  from /tmp/src/lavinmq/client/channel.cr:483:9 in '??'
  from /tmp/src/lavinmq/client/channel.cr:470:11 in 'basic_ack'
  from /tmp/src/lavinmq/client/client.cr:367:31 in 'process_frame'
  from /tmp/src/lavinmq/client/client.cr:125:13 in 'read_loop'
  from /usr/share/crystal/src/fiber.cr:146:11 in 'run'
  from ???
2024-04-26 10:12:31 +00:00 [INFO] client[vhost=/ address=10.42.0.160:42596] - Closing, INTERNAL_ERROR - Unexpected error, please report
```
Something seems to have gone wrong.
The service was restarted once before the queue was closed
@viktorerlingsson

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+### Fixed
+
+- Queues will no longer be closed if file size is incorrect. Fixes [#669](https://github.com/cloudamqp/lavinmq/issues/669)
+
 ## [2.0.2] - 2024-11-25
 
 ### Fixed

--- a/spec/storage_spec.cr
+++ b/spec/storage_spec.cr
@@ -99,7 +99,7 @@ describe LavinMQ::AMQP::DurableQueue do
     end
   end
 
-  it "handles files with few extra bytes" do
+  it "shift? handles files with few extra bytes" do
     queue_name = Random::Secure.hex(10)
     with_amqp_server do |s|
       vhost = s.vhosts.create("test_vhost")
@@ -125,6 +125,41 @@ describe LavinMQ::AMQP::DurableQueue do
         # read messages, should not raise any error
         q.subscribe(tag: "tag", no_ack: false, &.ack)
         should_eventually(be_true) { queue.empty? }
+      end
+    end
+  end
+
+  it "first? handles files with few extra bytes" do
+    queue_name = Random::Secure.hex(10)
+    with_amqp_server do |s|
+      vhost = s.vhosts.create("test_vhost")
+      with_channel(s, vhost: vhost.name) do |ch|
+        q = ch.queue(queue_name)
+        queue = vhost.queues[queue_name].as(LavinMQ::AMQP::DurableQueue)
+        mfile = queue.@msg_store.@segments.first_value
+
+        # fill up one segment
+        message_size = 41
+        while mfile.size < (LavinMQ::Config.instance.segment_size - message_size*2)
+          q.publish_confirm "a"
+        end
+        remaining_size = LavinMQ::Config.instance.segment_size - mfile.size - message_size
+        q.publish_confirm "a" * remaining_size
+
+        # publish one more message to create a new segment
+        q.publish_confirm "a"
+
+        # resize first segment to LavinMQ::Config.instance.segment_size
+        mfile.resize(LavinMQ::Config.instance.segment_size)
+
+        store = LavinMQ::Queue::MessageStore.new(queue.@msg_store.@queue_data_dir, nil)
+        mfile = store.@segments.first_value
+        mfile.pos = mfile.size - 2
+        if msg = store.first?
+          msg.@segment_position.@segment.should eq 2
+        else
+          fail "no message"
+        end
       end
     end
   end

--- a/src/lavinmq/amqp/queue/message_store.cr
+++ b/src/lavinmq/amqp/queue/message_store.cr
@@ -62,7 +62,7 @@ module LavinMQ
         notify_empty(false) if was_empty
       end
 
-      def first? : Envelope?
+      def first? : Envelope? # ameba:disable Metrics/CyclomaticComplexity
         raise ClosedError.new if @closed
         if sp = @requeued.first?
           seg = @segments[sp.segment]

--- a/src/lavinmq/amqp/queue/message_store.cr
+++ b/src/lavinmq/amqp/queue/message_store.cr
@@ -90,6 +90,11 @@ module LavinMQ
           msg = BytesMessage.from_bytes(rfile.to_slice + pos)
           sp = SegmentPosition.make(seg, pos, msg)
           return Envelope.new(sp, msg, redelivered: false)
+        rescue ex : IndexError
+          @log.warn { "Msg file size does not match expected value, moving on to next segment" }
+          select_next_read_segment && next
+          return if @size.zero?
+          raise Error.new(@rfile, cause: ex)
         rescue ex
           raise Error.new(@rfile, cause: ex)
         end

